anthos-bm-gcp-terraform/test/unit/module_vm_test.go

```diff
@@ -0,0 +1,372 @@
+		"Variable not found in plan: instance_template",
+	)
+	util.ExitIf(hasVar, false)
+
+	// verify input variable region in plan matches
+	assert.Equal(
+		goTester,
+		region,
+		vmInstancePlan.Variables.Region.Value,
+		"Variable does not match in plan: region.",
+	)
+
+	// verify input variable network in plan matches
+	assert.Equal(
+		goTester,
+		network,
+		vmInstancePlan.Variables.Network.Value,
+		"Variable does not match in plan: network.",
+	)
+
+	// verify input variable instance_template in plan matches
+	assert.Equal(
+		goTester,
+		instanceTemplate,
+		vmInstancePlan.Variables.InstanceTemplate.Value,
+		"Variable does not match in plan: instance_template.",
+	)
+
+	// verify size of input variable vm_names array in plan
+	assert.Len(
+		goTester,
+		vmInstancePlan.Variables.Names.Value,
+		len(expectedVMNames),
+		"Variable count does not match in plan: vm_names.",
+	)
+
+	// verify each input variable vm_name in plan matches
+	for _, vmName := range vmInstancePlan.Variables.Names.Value {
+		assert.Contains(
+			goTester,
+			expectedVMNames,
+			vmName,
+			"Variable does not match in plan: vm_names.",
+		)
+	}
+
+	// verify the number of resources planned
+	assert.Len(
+		goTester,
+		vmInstancePlan.PlannedValues.RootModule.ChildModules,
+		len(expectedVMNames)+1, // +1 for the external Ip resource
+		"Resource count does not match in plan: google_compute_address.",
+	)
+
+	numberOfComputeInstanceModules := 0
+	for idx, childModule := range vmInstancePlan.PlannedValues.RootModule.ChildModules {
+		moduleAddress := childModule.ModuleAddress
+		if strings.HasPrefix(moduleAddress, "module.compute_instance") {
+			numberOfComputeInstanceModules++
+			validateComputeInstanceSubModule(
+				goTester, &childModule, idx,
+				&expectedVMNames, instanceTemplate, network, region)
+
+		} else if strings.HasPrefix(moduleAddress, "module.external_ip_addresses") {
+			assert.Len(
+				goTester,
+				childModule.Resources,
+				len(expectedVMNames),
+				fmt.Sprintf("Invalid count for planned_values.root_module.child_modules[%d].resources", idx),
+			)
+
+			for ipIdx, externalIPResource := range childModule.Resources {
+				validateExternalIPInSubModule(
+					goTester, &externalIPResource, idx, ipIdx, &expectedVMNames, region)
+			}
+		} else {
+			// child module should be either compute_instance or external_ip_addresses
+			// it cannot be anything else
+			goTester.Error(fmt.Sprintf(
+				"Module 'vm' can only have the following sub modules: "+
+					"\n\t- compute_instance"+
+					"\n\t- external_ip_addresses"+
+					"\n But it also has: %s",
+				moduleAddress,
+			))
+			goTester.Fail()
+		}
+	}
+	// verify how many child modules of type google_compute_instance_from_template were present
+	assert.Len(
+		goTester,
+		expectedVMNames,
+		numberOfComputeInstanceModules,
+		"Resource count for module type google_compute_instance_from_template does not match in plan",
+	)
+
+	validateOutputs(goTester, vmInstancePlan.PlannedValues.Outputs)
```

Comment:
Generic comment: Will these tests (e.g. validate compute instance sub module) work when run on say other cloud provider or on-prem/disconnected?

Reply:
Yes - as long as we can keep the same name for `input & output` variables for the module.